engine/engine/urls.py

```diff
@@ -68,3 +73,7 @@
     urlpatterns += [path(settings.SILK_PATH, include("silk.urls", namespace="silk"))]
 
 admin.site.site_header = settings.ADMIN_SITE_HEADER
+
+# NOTE: keep this at the end of the file to properly override urlpatterns
+if settings.IS_IN_MAINTENANCE_MODE:
+    urlpatterns = paths_to_work_even_when_maintenance_mode_is_active
```

Comment:
@mderynck raised a good point here about this that I agree with. Allowing traffic to these routes is less deterministic when the database is down. Some routes might work, some might complete a small amount of work before returning a 500.

Reply:
Maybe we need to check the setting when adding middleware too, add a custom middleware to return 503 for endpoints that would resolve normally. Returning 404 is bad because people are more likely to fiddle with their config if they didn't know we are doing maintenance (Did not pay attention to the warning).  503 at least lets them know its on our side.